xcode/Subconscious/Shared/Services/NoosphereService.swift

```diff
@@ -0,0 +1,204 @@
+    case foreignError(String)
+    case memoDoesNotExist
+}
+
+public struct SphereReceipt {
+    var identity: String
+    var mnemonic: String
+}
+
+struct SphereMemo: Hashable {
+    var contentType: String
+    var data: Data
+}
+
+/// Create a Noosphere instance.
+///
+/// - Property noosphere: pointer that holds all the internal book keeping.
+///   DB pointers, key storage interfaces, active HTTP clients etc.
+public final class Noosphere {
+    private let noosphere: OpaquePointer
+    
+    init(
+        globalStoragePath: String,
+        sphereStoragePath: String
+    ) throws {
+        guard let noosphere = ns_initialize(
+            globalStoragePath,
+            sphereStoragePath,
+            nil
+        ) else {
+            throw NoosphereError.foreignError(
+                "Failed to get pointer for Noosphere"
+            )
+        }
+        self.noosphere = noosphere
+    }
+    
+    /// Create and configure a user and sphere
+    /// - Important: the identity should be persisted
+    /// - Important: the mnemonic should be displayed to user and then discarded.
+    /// - Returns: a SphereConfig containing the sphere identity and mnemonic
+    ///
+    /// What it does:
+    /// - Initializes a namespace
+    /// - Creates a key
+    /// - Creates a sphere for that key
+    public func createSphere(
+        ownerKeyName: String
+    ) throws -> SphereReceipt {
+        ns_key_create(noosphere, ownerKeyName)
+        
+        guard let sphereReceipt = ns_sphere_create(
+            noosphere,
+            ownerKeyName
+        ) else {
+            throw NoosphereError.foreignError(
+                "Failed to get pointer for sphere receipt"
+            )
+        }
+        defer {
+            ns_sphere_receipt_free(sphereReceipt)
+        }
+        
+        guard let sphereIdentityPointer = ns_sphere_receipt_identity(
+            sphereReceipt
+        ) else {
+            throw NoosphereError.foreignError(
+                "Failed to get pointer for identity"
+            )
+        }
+        defer {
+            ns_string_free(sphereIdentityPointer)
+        }
+        
+        guard let sphereMnemonicPointer = ns_sphere_receipt_mnemonic(
+            sphereReceipt
+        ) else {
+            throw NoosphereError.foreignError(
+                "Failed to get pointer for mnemonic"
+            )
+        }
+        defer {
+            ns_string_free(sphereMnemonicPointer)
+        }
+        
+        let sphereIdentity = String.init(cString: sphereIdentityPointer)
+        let sphereMnemonic = String.init(cString: sphereMnemonicPointer)
+        
+        return SphereReceipt(
+            identity: sphereIdentity,
+            mnemonic: sphereMnemonic
+        )
+    }
+
+    /// Perform an action with SphereFS using a closure.
+    /// SphereFS pointer is freed after action is performed.
+    /// - Returns: return value of `action` closure
+    @discardableResult private func withSphereFS<T>(
```

Comment:
Helper that opens the sphere file system, executes closure, passing a pointer to the fs, then frees up the sphere fs memory on complete.